### ISSUE TITLE
Consider making all Flax patterns "vmap-first"

### ISSUE THREAD

**Issue by avital** [google/flax#690]
I believe that most of our current layers work regardless of whether a batch dimension is present. We could change our best practices to vmap over the batch dimension at the top-level of training code, or alternatively, wrapping only your top-level module with vmap rather than making our built-in modules designed around batches.

**Comment by andsteing** [comment]
So do you think we should then also rewrite all the examples to operate on single examples up and then lift the main model `Module` ?

**Comment by avital** [comment]
We should first understand the consequences, but yes that seems like a reasonable path. I think @levskaya had some more nuanced thoughts about what it would take to get to a vmap-first future.

**Comment by jheek** [comment]
I think we should be careful here as well. VMap does not always make things clearer to understand. It definitely hurts inspectability of modules because the modules inside are no longer vectorized so you would have to re-vectorize them. For example when you use just the decoder from a transformer.

Also the Jax team might still make some major changes with vmap/pmap being replaced by a generic map and collectives being replaced by named axes so `jax.lax.psum(x, 'batch')` becomes `jax.numpy.sum(x, 'batch')`. With generic map it might be more useful because it would mean you get multi device training "for free"

**Comment by levskaya** [comment]
I think we should make sure that we support "vmap first" in the sense that the core layers allow for single-example layers/models.  eg. adding our fix to Conv that allows single-example, and maybe a pass-through for batchnorm? 

I agree w. Jonathan that I'm not yet convinced we should switch all the models and examples over to this style yet though, esp. given the jax changes coming up, though perhaps we could make a small example showing how you'd do it if you wanted to write models this way.

**Comment by levskaya** [comment]
Closing as  #740 #564 resolved this issue as far as I understand.
